app/src/interfaces/_system/system-permissions/system-permissions.vue

```diff
@@ -0,0 +1,714 @@
+
+			const deleteIndex = _value.value.delete.findIndex((id) => id === item[targetPKField]);
+
+			if (deleteIndex !== -1) {
+				merge(updatedItem, { $type: 'deleted', $index: deleteIndex });
+			}
+
+			return updatedItem;
+		});
+
+		items.push(...createdItems.value);
+
+		if (addMinimalAppPermissions.value) {
+			for (const perm of appAccessMinimalPermissions) {
+				if (items.find((item) => item.collection === perm.collection && item.action === perm.collection)) continue;
+				items.push(perm);
+			}
+		}
+
+		const sortField = 'collection';
+
+		return items.sort((a, b) => {
+			return get(a, sortField) - get(b, sortField);
+		});
+	});
+
+	const { cleanItem, isEmpty, isLocalItem, getItemEdits } = useUtil();
+	const { create, update, remove } = useActions(_value);
+	const { resetActive, resetSystemPermissions } = useReset();
+
+	return {
+		loading,
+		create,
+		update,
+		remove,
+		displayItems,
+		isLocalItem,
+		getItemEdits,
+		resetActive,
+		resetSystemPermissions,
+	};
+
+	function useActions(target: Ref<any>) {
+		return { create, update, remove };
+
+		function create(...items: Record<string, any>[]) {
+			for (const item of items) {
+				target.value.create.push(cleanItem(item));
+			}
+
+			updateValue();
+		}
+
+		function update(...items: DisplayItem[]) {
+			if (!relation.value) return;
+
+			for (const item of items) {
+				if (item.$type === undefined || item.$index === undefined) {
+					target.value.update.push(cleanItem(item));
+				} else if (item.$type === 'created') {
+					target.value.create[item.$index] = cleanItem(item);
+				} else if (item.$type === 'updated') {
+					if (isEmpty(item)) target.value.update.splice(item.$index, 1);
+					else target.value.update[item.$index] = cleanItem(item);
+				} else if (item.$type === 'deleted') {
+					if (item.$edits !== undefined) {
+						if (isEmpty(item)) target.value.update.splice(item.$index, 1);
+						else target.value.update[item.$edits] = cleanItem(item);
+					} else {
+						target.value.update.push(cleanItem(item));
+					}
+
+					target.value.delete.splice(item.$index, 1);
+				}
+			}
+
+			updateValue();
+		}
+
+		function remove(...items: DisplayItem[]) {
+			if (!relation.value) return;
+
+			// Ensure that the objects are removed in reverse order to not mess up the indices while we're working with them
+			items = orderBy(items, ['$index'], ['desc']);
+
+			for (const item of items) {
+				if (item.$type === undefined || item.$index === undefined) {
+					target.value.delete.push(item.id);
+				} else if (item.$type === 'created') {
+					target.value.create.splice(item.$index, 1);
+				} else if (item.$type === 'updated') {
+					target.value.update.splice(item.$index, 1);
+					target.value.delete.push(item.id);
+				} else if (item.$type === 'deleted') {
+					target.value.delete.splice(item.$index, 1);
+				}
+			}
+
+			updateValue();
+		}
+
+		function updateValue() {
+			target.value = cloneDeep(target.value);
+		}
+	}
+
+	function useUtil() {
```

Comment:
Wondering if we should extract some of these useX functions to separate ts files to organize this massive setup function a bit